internal/types/subtypes/interceptor.go

```diff
@@ -0,0 +1,193 @@
+package subtypes
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/sdk/pbs/controller/protooptions"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+func messageDomain(m proto.Message) string {
+	r := m.ProtoReflect()
+	fd := r.Descriptor().ParentFile()
+	if fd == nil {
+		return ""
+	}
+
+	opts, ok := fd.Options().(*descriptorpb.FileOptions)
+	if !ok {
+		return ""
+	}
+
+	domain := proto.GetExtension(opts, protooptions.E_Domain).(string)
+	return domain
+}
+
+func transformRequestAttributes(_ context.Context, req proto.Message) proto.Message {
+	domain := messageDomain(req)
+
+	r := req.ProtoReflect()
+	fields := r.Descriptor().Fields()
+
+	itemField := fields.ByName("item")
+	idField := fields.ByName("id")
+	attributesField := fields.ByName("attributes")
+
+	fieldValue := func(m protoreflect.Message, fd protoreflect.FieldDescriptor) string {
+		if fd == nil {
+			return ""
+		}
+		return m.Get(fd).String()
+	}
+
+	var st Subtype
+	switch {
+	case itemField != nil:
```

Comment:
Could we add a comment explaining the scenario that this implies and the next steps? E.g.
```
If "item" is set, it implies that we're sending a resource in the request, e.g. via a Create call.
Use the following priority for deducing the subtype:
* Look it up from the "id" field, if set.
* Look it up from any specified source ID field.
* Look it up by from the "type" field,
```

I think this code is pretty self documenting, but I'm worried it's because we've spent weeks in reflect land already, and whenever reflection is concerned, it's never a bad idea to over-comment, IMO.

Reply:
I can look to add some additional comments in this file. I went back and forth on this a bit. I agree that it is a bit "self documenting", and I think the corresponding tests also provide some good context. The only down-side I have see to over-commenting is when the code gets updated and the comments dont.

Reply:
Yeah, agree, hopefully this code won't change too much

Reply:
See [7658e9e](https://github.com/hashicorp/boundary/pull/1967/commits/7658e9e489454006241899d5864127cb99e47422)